Initial implementation of service

mirdono opened the pull request:
This service provides functionality to manage the location resources that can be used scope of operation (*nl. werkingsgebied*) for administrative units. Due to constraints in the data model[^1] an administrative unit can only be linked to at most 1 location as its scope of operation. Consequently, an appropriate location resource may have to be created whenever an administrative unit is assigned multiple locations as its scope of operation.

In summary this service provides three functions
1. Determine the label that should be shown in the user for a given location. This is either simply the label directly linked to the location resource or a concatenation of the labels for all locations contained in the given one.
2. Determine all location resources that are contained within a given one.
3. Find or, if necessary, create a location resource that contains a given set of other location resources.

## Points of discussion
### Level for new location resources
In OP location resources have a level[^2], the currently existing locations have as values "District", "Gemeente", "Referentieregio", or "Province". It remains to be determined what level a newly created location resource should receive. The question was put to business but they had no opinion on it.

Note, that reusing any of the existing levels likely impacts the current implementation of this service. For example, the expected functionality is to show users a list of municipality-level locations as a unit's scope of operation. Therefore, for units linked with a "Referentieregio" level location, this service looks for the contained locations and uses their labels instead of the label of the directly linked location resource.

### Writing to public graph
In preparation of this functionality all location data was moved to the public graph, see [#542](https://github.com/lblod/app-organization-portal/pull/542). Therefore, this service also writes any newly created location resources to the `public` graph. This has the advantage that it keeps all location data in a single graph. This avoids additional complexity in the configuration of other services such as the producers and authorisation layer.

## Notes
- As part of the MVP locations at the "District" level are ignored. They are only relevant for the municipality of Antwerp and may be included in a later version.

## Related PRs
- Backend: [#550](https://github.com/lblod/app-organization-portal/pull/550)
- Frontend: [#680](https://github.com/lblod/frontend-organization-portal/pull/680)

## Related tickets
- OP-3205

[^1]: See the definition of an [administrative unit](https://data.vlaanderen.be/doc/applicatieprofiel/besluit-publicatie/#Bestuurseenheid) in the Besluit application profile.
[^2]: Using the self-defined predicate `<http://mu.semte.ch/vocabularies/ext/werkingsgebiedNiveau>`

mirdono committed:
feat: get label to display for an organisation's scope of operation

Functionality to retrieve the label that should be displayed for an
organisation's scope of operation. This is usually just the label linked to the
location resources. Except for scopes of the levels reference
region (*nl. Referentieregio*) or province. For such scopes, the label to that
should be displayed is the concatenation of the labels of all contained
municipality locations.

mirdono committed:
feat: retrieve locations within an organisation's scope of operation

Functionality to map an organisation's scope of operation to the location(s)
that are contained in it. For locations of the municipality level this is the
location itself as, for the MVP, a municipality location is considered to be the
smallest granularity. For other levels of locations, the data is queried to
retrieve all locations that are `geo:sfwithin` the organisation's scope.

mirdono committed:
fix: use multiple queries to determine scope for a set of locations

The single query initially used to retrieve a location resource that contains a
given set of location resources only worked for small sets of contained
locations. Once the set of contained locations became to large that query caused
several problems, up to crashing the triplestore.

As a solution this flow is split over two queries, which are glued together by
comparing sets of included locations within the service. A first query retrieves
the candidate location resources that contain the right amount of
locations. From these candidates one is sought whose actually contained
locations, retrieved by a second query, match exactly the intended set.

mirdono committed:
chore: add readme

mirdono committed:
chore: silence warning about unused argument

mirdono committed:
feat: create scope for a given set of locations

An organisation can operate in multiple locations, but the datamodel allows an
administrative unit to be linked to at most one location as its scope of
operation.
Therefore, this service will, when necessary, create new location resources that
can be linked to organisations.

claire-lovisa commented on README.md:
```diff
@@ -0,0 +1,49 @@
+- `404 No Content` No matching location resource or label was found.
+- `500 Internal Server Error` Something went wrong, check the service logs for more details.
+
+### `GET locations-in-scope/:locationUuid`
```
I'm wondering if we shouldn't return URIs instead of UUIDs. In theory, they are the true reference that will never be changed. In practice, UUIDs don't change much and the result of this call will be used instantly by the calling app. I'll leave the decision to you haha

mirdono replied:
I have gone back and forth on this myself. I ended up settling on returning UUIDs as that is what the OP frontend currently tends to use in calls to the backend. This way the frontend can just do its thing with the response as usual, no need to deviate from the rest its code base by using calls with URIs or having to extract UUIDs from the received URIs. I am not sure if I a breaking any conventions or best practices here.

claire-lovisa replied:
I'm a bit unsure as well. I think for this context that UUID is a decent choice and if OP uses mostly UUIDs overall, i think it's a good argument.

claire-lovisa commented on app.js:
```diff
@@ -1,8 +1,201 @@
 
-import { app, query, errorHandler } from 'mu';
-app.get("/", function (req, res) {
+/** @import { LocationDetails } from "./lib/queries"; */
```
This comment can go away?

mirdono replied:
I know it looks a bit odd, but it does have its purpose. I use the [TS language server](https://github.com/typescript-language-server/typescript-language-server) in my JS/TS development setup. In JS this can use the type information in the jsdoc to annotate the code with type information. For that to work correctly, in this case, the `LocationDetails` typedef (defined in the `queries` file) needs to be imported here, which is what this comment does[^1]. Without it the language server will be complain that it does not know that type.
(Side note, I considered to write this service fully in TS as some other teams started doing recently. But decided against it because I was a bit worried that that would add additional complexity.)

[^1]: https://devblogs.microsoft.com/typescript/announcing-typescript-5-5-rc/#type-imports-in-jsdoc

claire-lovisa commented on lib/util.js:
```diff
@@ -0,0 +1,51 @@
+/**
+ * Check whether a location's label needs to be the aggregation of the its
+ * contained locations.
+ * @param {...import("./lib/queries").LocationDetails} location - The location
```
Should we remove the `...import("./lib/queries").` part and only keep `LocationDetails`?

mirdono replied:
Yes, correct. The import part is not anymore due to the `@import` comment near the top of the file.

claire-lovisa commented on config.js:
```diff
@@ -0,0 +1,14 @@
+  rdfs: "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>",
+};
+
+// TODO: Should we write these to public? The advantage is that they than all
```
I'm quoting you here, you said in the PR https://github.com/lblod/app-organization-portal/pull/542 that "the existing locations are essentially a "standard" set that should not change over time". I get that those extra locations, that are a bit "non-standard" feel off to add there too but I think it makes sense to keep all locations in the same bucket. Could we maybe think of a way of differentiating those "new" locations from the standard ones? Maybe via a comment stored in the db, or a mention that it has been created by this service?

mirdono replied:
I agree, that there are good arguments for inserting these new locations in the `public` graph. I just wanted to explicitly raise this point due to the fact that it feels a bit off and wanted to make sure it is a reasonable thing to do.

Differentiating the new locations from the "standard" ones could be handy.

claire-lovisa commented on app.js:
```diff
@@ -1,8 +1,201 @@
+  const label = getSortedLabels(...containedLocations);
+
+  // TODO: determine level, requires input from business
+  const level = "TODO";
```
Will you get an answer at some point, or was their answer "we have no opinion on it" as you said in the PR description?
In the meantime we could use something vague, "Concatenated locations" or something like that, a bit more descriptive than TODO haha

mirdono replied:
The answer was basically, "this seems something for the data modelling team".  I raise it here first, but if we don't reach a decision I'll check on the `lblod-all` channel for opinions. A new level specific for these locations would be easiest, but I am not sure if we can just add that without causing issues somewhere down the line.

claire-lovisa reviewed (approved):
The service seems to work as promised while testing so I approve, but I had some thoughts while reading the code so I'll let you check it out :)

mirdono committed:
feat: identify location resources created by the service

Extend the `prov:Location` resources created by this service with a
`dcterms:creator` triple to identify them as having been created by this
service. This allows to differentiate the ad-hoc resources created by this
service from the, somewhat, standard set of scope of operation locations that is
already defined.
